File: rust/src/s0200.rs
pub struct Solution {}

impl Solution {
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let mut result = 0;
        let (m, n) = (grid.len(), grid[0].len());
        for i in 0..m {
            for j in 0..n {
                if grid[i][j] == '1' {
                    result += 1;
                    Self::dfs(&mut grid, i, j);
                }
            }
        }
        result
    }

    fn dfs(grid: &mut Vec<Vec<char>>, i: usize, j: usize) {
        let (m, n) = (grid.len(), grid[0].len());
        if i >= m || j >= n {
            return;
        }
        if grid[i][j] == '0' {
            return;
        }
        grid[i][j] = '0';
        Self::dfs(grid, i + 1, j);
        Self::dfs(grid, i, j + 1);
        if i > 0 {
            Self::dfs(grid, i - 1, j);
        }
        if j > 0 {
            Self::dfs(grid, i, j - 1);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        Solution {};
    }
}

File: cpp/s0200.cc
class Solution {
public:
  int numIslands(vector<vector<char>> &grid) {
    int result = 0;
    for (int m = 0; m < grid.size(); m++) {
      for (int n = 0; n < grid[0].size(); n++) {
        if (grid[m][n] == '1') {
          result += 1;
          dfs(grid, m, n);
        }
      }
    }
    return result;
  }
  void dfs(vector<vector<char>> &grid, int m, int n) {
    if (m < 0 || m >= grid.size() || n < 0 || n >= grid[0].size()) return; // edge case
    if (grid[m][n] == '0') return; // nothing further
    // mark as visited
    grid[m][n] = '0';
    // traverse 4 way up, down, left right
    dfs(grid, m, n-1);
    dfs(grid, m, n+1);
    dfs(grid, m-1, n);
    dfs(grid, m+1, n);
  }
};
